### ISSUE TITLE
Using .coverage from pytest-cov says "Nothing to report"

### ISSUE THREAD

**Issue by AbdealiLoKo** [lemurheavy/coveralls-public#1737]
Earlier I was creating lcov files and uploading lcov files.

As was suggested in https://github.com/lemurheavy/coveralls-public/issues/1736#issuecomment-1806150754 - I tried using the .coverage database directly with coveralls.

But that does not seem to be working:
![image](https://github.com/lemurheavy/coveralls-public/assets/2200743/3a4c5f74-0683-4e3f-bc75-0062c2ef22c8)

I can see that the `.coverage` file has been created. And it has content - if I open `sqlite3 .coverage` and so `select * from file` it shows me the source code files that coverage has found and measured.
The console summary from pytest-cov also shows that coverage was captured fine.

I noticed that `pytest-cov` was marked as beta status [here](https://github.com/coverallsapp/coverage-reporter#supported-coverage-report-formats). and it says:
> Parsing is done in a pretty naive way
So, is there some logic where if coveralls is unable to find the source code - it will ignore coverage from that file ?

I have a monorepo setup. Where I create my projects under:
```
apps/
    .coveragerc
    project-1
        project_1
        test
    project-2
        project_2
        test
```
My `.coverage` file contains:
```
$ sqlite3 .coverage
sqlite> select * from file;
1|/home/abdealiloko/myrepo/apps/package-1/package_1/__init__.py
2|/home/abdealiloko/myrepo/apps/package-1/package_1/abc.py
```

And my `.coveragerc` file is:
```
[run]
dynamic_context = test_function
branch = true
cover_pylib = False
include = apps/**
```

**Comment by afinetooth** [comment]
Hi, @AbdealiLoKo.

Have you tried the recommendation from the documentation linked from that ["beta" link](https://github.com/coverallsapp/coverage-reporter#supported-coverage-report-formats) next to `pytest-cov`? It links to [more info about that](https://github.com/coverallsapp/coverage-reporter/#pytest-cov-test_tube-beta) and suggests exporting to `XML`, per:

> #### Pytest-Cov (🧪 beta)
> Since .coverage stores only covered lines coverage-reporter needs to parse Python code to get uncovered lines. Parsing is done in a pretty naive way, so the results might be different from actual especially for complicated and non-trivial code.
> 
> If coverage results are incorrect consider exporting .coverage to XML:
> 
> ```
> coverage xml # creates coverage.xml
> coveralls -f coverage.xml
> ```

I have asked our engineering team about how well that is supported, as I don't see if specifically in the source code. I only see [the code that parses the `.sqlite` file](https://github.com/coverallsapp/coverage-reporter/blob/master/src/coverage_reporter/parsers/coveragepy_parser.cr#L33-L62).

**Comment by afinetooth** [comment]
@AbdealiLoKo also, can you please share two things:

1. The Coveralls URL to your Coveralls build that showed "Nothing to report."; and
2. If possible, the original `.coverage` file, if you happened to store it as an artifact in CI.

Thanks.

**Comment by AbdealiLoKo** [comment]
With lcov file - it works perfectly fine. I have not tried it with XML conversion

I can't seem to find the original build where this happened. Will try to reproduce and send an email
We can add this to the issues @indiVar0508 is tracking with the support email ?

**Comment by AbdealiLoKo** [comment]
Nevermind - I found the build, It was in my public coveralls test repo:
https://github.com/AbdealiLoKo/coveralls-issue/actions/runs/6847690062
The coverage file is present in the artifacts.

**Comment by afinetooth** [comment]
Hi, @AbdealiLoKo.

> With lcov file - it works perfectly fine. I have not tried it with XML conversion

Either of those formats will work fine, but you will have the best results if you add the [`format` input option](https://github.com/marketplace/actions/coveralls-github-action#:~:text=format,formats%20list.) to your upload step(s), so:

```
      - name: Send coverage data to Coveralls
        uses: coverallsapp/github-action@v2
        with:
          flag-name: lcov-myapp2
          parallel: true
          format: cobertura # for XML
```

For `xml` format. (`coverage.py`'s XML export is in Cobertura format.)

Or:

```
      - name: Send coverage data to Coveralls
        uses: coverallsapp/github-action@v2
        with:
          flag-name: lcov-myapp2
          parallel: true
          format: lcov # for LCOV
```

For `lcov` format.

This is particularly true if you're exporting _multiple_ identical coverage reports in two or more formats. It will just avoid any corner cases for converting and merging across formats.

> We can add this to the issues @indiVar0508 is tracking with the support email ?

Of course.

> Nevermind - I found the build, It was in my public coveralls test repo:
> https://github.com/AbdealiLoKo/coveralls-issue/actions/runs/6847690062
> The coverage file is present in the artifacts.

OK, thank you.

## tl;dr:
I was able to confirm that Coverage Reporter, whose parser for `coverage.py` is still in beta, is not processing any of the branch coverage (`arcs`) in your `.coverage` files once it determines that there is no line coverage data (`line_bits`) in them.

### Solution:
Continue exporting your python coverage reports in `LCOV` or `XML` format for now, while we modify our `coverage.py` parser. 

**Note**: _As an aside, due to some variance in the format of the SQLite database across versions, like other tools we will probably punt on translating the `.coverage` file and ultimately convert it into (Cobertura) `XML` format, so, for now, you can achieve the same outcome by converting to `XML` (or `LCOV`)._

You can continue doing that, for `xml` with:

```
venv/bin/pytest \
  -k "test_text_order_by" \
    --cov --cov-report=xml --cov-report=html
```

Or, for `lcov`:

```
venv/bin/pytest \
  -k "test_text_order_by" \
    --cov --cov-report=lcov --cov-report=html
```

But, again, you'll get better results by specifying your format, like:

```
      - name: Send coverage data to Coveralls
        uses: coverallsapp/github-action@v2
        with:
          flag-name: lcov-myapp2
          parallel: true
          format: cobertura # for XML
```

Or:

```
      - name: Send coverage data to Coveralls
        uses: coverallsapp/github-action@v2
        with:
          flag-name: lcov-myapp2
          parallel: true
          format: lcov # for LCOV
```

## Details:
I looked at the resulting (SQLite-format) `.coverage` files from `test-myapp1` and `test-myapp2` and discovered why we are returning "Nothing to report" from the Coverage Reporter (the integration underlying the Coveralls GitHub Action),

It's because we make an initial query for line coverage against the `line_bits` SQLite database table and, at least in the two `.coverage` files from your [artifacts here](https://github.com/AbdealiLoKo/coveralls-issue/actions/runs/6847690062), that table is empty.

When I explore the (SQLite) `.coverage`file for `test-myapp1` I see the following:

```
% pwd              
/Users/.../Downloads/CORRIDOR/myapp1-coverage-data/test_results/myapp
% sqlite3 .coverage
SQLite version 3.39.5 2022-10-14 20:58:05
Enter ".help" for usage hints.
sqlite> .tables
arc              coverage_schema  line_bits        tracer         
context          file             meta  

sqlite> SELECT * FROM arc;
1,1,-1,1
1,1,3,-1
1,1,1,3
2,1,-1,1
[...]
3,2,31,37
4,2,-23,25
4,2,25,26
4,2,26,27
4,2,27,28
4,2,28,-23

sqlite> SELECT * FROM context;
1,""
2,test_main.test_duplicates_in_order_by

sqlite> SELECT * FROM coverage_schema;
7

sqlite> SELECT * FROM file;
1,/home/runner/work/coveralls-issue/coveralls-issue/myapp_test/conftest.py
2,/home/runner/work/coveralls-issue/coveralls-issue/myapp/__init__.py
3,/home/runner/work/coveralls-issue/coveralls-issue/myapp/main.py
4,/home/runner/work/coveralls-issue/coveralls-issue/myapp_test/test_main.py

sqlite> SELECT * FROM line_bits;

sqlite> SELECT * FROM meta;
version,7.3.2
has_arcs,1

sqlite> SELECT * FROM tracer;
1,""
2,""
3,""
4,""
```

When I replicate the first [query](https://github.com/coverallsapp/coverage-reporter/blob/master/src/coverage_reporter/parsers/coveragepy_parser.cr#L11-L13) our (beta) [parser](https://github.com/coverallsapp/coverage-reporter/blob/master/src/coverage_reporter/parsers/coveragepy_parser.cr) makes, I get this result:

```
sqlite> SELECT file.path, line_bits.numbits FROM line_bits INNER JOIN file ON (line_bits.file_id = file.id);

```

In other words, nothing.

Similar results for `test-myapp2`:

```
% pwd
/Users/.../Downloads/CORRIDOR/myapp2-coverage-data/test_results/myapp
% sqlite3 .coverage      
SQLite version 3.39.5 2022-10-14 20:58:05
Enter ".help" for usage hints.
sqlite> .tables
arc              coverage_schema  line_bits        tracer         
context          file             meta           

sqlite> SELECT * FROM arc;
1|1|-1|1
1|1|3|-1
1|1|1|3
2|1|-1|1
[...]
3|2|18|-12
4|2|-15|17
4|2|17|18
4|2|18|19
4|2|19|20
4|2|20|-15

sqlite> SELECT * FROM context;
1|
2|test_main.test_text_order_by

sqlite> SELECT * FROM coverage_schema;
7

sqlite> SELECT * FROM file;
1|/home/runner/work/coveralls-issue/coveralls-issue/myapp_test/conftest.py
2|/home/runner/work/coveralls-issue/coveralls-issue/myapp/__init__.py
3|/home/runner/work/coveralls-issue/coveralls-issue/myapp/main.py
4|/home/runner/work/coveralls-issue/coveralls-issue/myapp_test/test_main.py

sqlite> SELECT * FROM line_bits;

sqlite> SELECT * FROM meta;
version|7.3.2
has_arcs|1

sqlite> SELECT * FROM tracer;
1|
2|
3|
4|

sqlite> SELECT file.path, line_bits.numbits FROM line_bits INNER JOIN file ON (line_bits.file_id = file.id);

```

Again, nothing. 

I do find this strange, since the `LCOV` versions of the same reports _do_ apparently contain line coverage.

Here's an excerpt of the coverage report we received for [this build](https://coveralls.io/builds/63913484), when you were still exporting in `LCOV` format:

```
{
  "service_branch": "main",
  "service_build_url": "https://github.com/AbdealiLoKo/coveralls-issue/actions/runs/6847598414",
  [...]
  "source_files": [
    {
      "name": "myapp/__init__.py",
      "coverage": [
        1,
        null,
        1,
        null,
        1
      ],
      "source_digest": "d28290417432cf77af56bb3a272bde34"
    },
    {
      "name": "myapp/main.py",
      "coverage": [
        1,
        null,
        1,
        null,
        null,
        1,
        1,
        1,
        1,
        null,
        null,
        1,
        1,
        null,
        null,
        1,
        1,
        1,
        null,
        null,
        null,
        null,
        1,
        1,
        1,
        1,
        null,
        null,
        null,
        0,
        1,
        1,
        null,
        null,
        null,
        null,
        1,
        null,
        null,
        1,
        0
      ],
      "branches": [
        17,
        0,
        0,
        1,
        23,
        0,
        1,
        1,
        16,
        1,
        0,
        1,
        18,
        1,
        1,
        1,
        0,
        2,
        0,
        0,
        0,
        3,
        0,
        0,
        0,
        6,
        1,
        1,
        24,
        2,
        1,
        1,
        26,
        3,
        1,
        1,
        30,
        4,
        0,
        0,
        31,
        4,
        1,
        1,
        32,
        5,
        0,
        1,
        37,
        5,
        1,
        1,
        41,
        6,
        0,
        0
      ],
    },
    [...]
  ],
  "parallel": true,
  "git": {
    "branch": "main",
    "head": {
      [...]
    }
  },
  "run_at": "2023-11-13T08:46:06+00:00"
}
```

I don't know if you want to look into that, but it's not necessary if you continue exporting in `lcov` or `xml`.

**Comment by afinetooth** [comment]
Just to summarize for others reading for a quick solution:

# Solution
Continue exporting your python coverage reports in `LCOV` or `XML` format for now, while we modify our `coverage.py` parser (currently in beta). 

But you'll get the best results by specifying your `format`, like:

```
      - name: Send coverage data to Coveralls
        uses: coverallsapp/github-action@v2
        with:
          flag-name: lcov-myapp2
          parallel: true
          format: cobertura # for XML
```

Or:

```
      - name: Send coverage data to Coveralls
        uses: coverallsapp/github-action@v2
        with:
          flag-name: lcov-myapp2
          parallel: true
          format: lcov # for LCOV
```
